File: apps/pool-backend/src/category.py
"""Pool category (user intent) taxonomy.

`category` is the user intent stored on `lm_model` / derived models.json.
Physical modality is derived from intent — it is not a second CHECK column.

    chat, code  → text
    image       → image
    audio       → audio   (STT / TTS / SFX; e.g. Whisper, AudioGen)
    music       → audio   (text-to-music / song gen; same waveform domain)
    video       → video

`music` is not a third LLM intent beside chat/code. It is the generation
intent of the audio modality. Whisper-style STT stays `audio`; MusicGen /
MAGNeT / JASCO / ACE-Step / YuE / DiffRhythm / Stable Audio land as `music`.

Multimodal cells (audio + music) use RTF / ×realtime / wall / peak VRAM /
durationS — never decode_tok_s.
"""

from __future__ import annotations

import re

# Product intents accepted by the SQLite CHECK, API filter, and derive.
# `vision` is listed in the UI but has no category yet (no collection plan).
MODEL_CATEGORIES = ("chat", "code", "image", "audio", "music", "video")

TEXT_CATEGORIES = frozenset({"chat", "code"})
# Intents that share the audio waveform domain (STT/SFX vs song gen).
AUDIO_MODALITY_CATEGORIES = frozenset({"audio", "music"})
MULTIMODAL_CATEGORIES = frozenset({"image", "audio", "music", "video"})

INTENT_MODALITY = {
    "chat": "text",
    "code": "text",
    "image": "image",
    "audio": "audio",
    "music": "audio",
    "video": "video",
}

_MUSIC_RE = re.compile(
    r"musicgen|jasco|ace-?step|acestep|diffrhythm|stable[\s_-]?audio"
    r"|\bmagnet\b|magnet-|yue2|\byue\b|text-to-music|text2music",
    re.I,
)
_AUDIO_RE = re.compile(
    r"whisper|wav2vec|audiogen|faster-whisper|distil-whisper"
    r"|speech[\s_-]?to[\s_-]?text|\bstt\b|\btts\b",
    re.I,
)
_CODE_RE = re.compile(r"coder|starcoder|codestral|code", re.I)


def category_check_sql() -> str:
    """Literal CHECK tuple for the lm_model.category column."""
    inner = ",".join(f"'{name}'" for name in MODEL_CATEGORIES)
    return f"({inner})"


def modality_of(category: str) -> str | None:
    return INTENT_MODALITY.get(category)


def is_multimodal_category(category: str | None) -> bool:
    return category in MULTIMODAL_CATEGORIES


def validate_category_filter(category: str | None) -> str | None:
    """Return the filter value, or None for 'no filter'. Invalid → ValueError."""
    if category is None:
        return None
    if category not in MODEL_CATEGORIES:
        raise ValueError(f"invalid category {category!r}")
    return category


def model_category(display_name: str, hf_id: str = "") -> str:
    """Classify a catalog row into a pool intent.

    Music is matched before audio so MusicGen does not fall through to chat.
    Audio (Whisper / AudioGen) is matched before the legacy code/chat split
    so STT is never labelled as an LLM intent. Unknown names stay `chat`.
    """
    text = f"{display_name or ''} {hf_id or ''}"
    if _MUSIC_RE.search(text):
        return "music"
    if _AUDIO_RE.search(text):
        return "audio"
    if _CODE_RE.search(text):
        return "code"
    return "chat"

File: apps/web/scripts/category.mjs
// Pool category (user intent) taxonomy. Keep in lockstep with
// apps/pool-backend/src/category.py — tests assert the same names classify
// the same way on both sides.
//
// category is the user intent. Physical modality is derived:
//   chat, code  → text
//   image       → image
//   audio       → audio   (STT / TTS / SFX; e.g. Whisper, AudioGen)
//   music       → audio   (text-to-music / song gen; same waveform domain)
//   video       → video
//
// music is not a third LLM intent beside chat/code. It is the generation
// intent of the audio modality. Multimodal cells use RTF / ×realtime /
// wall / peak VRAM / durationS — never decode_tok_s.

export const MODEL_CATEGORIES = Object.freeze([
  "chat", "code", "image", "audio", "music", "video",
]);

export const TEXT_CATEGORIES = Object.freeze(new Set(["chat", "code"]));
export const AUDIO_MODALITY_CATEGORIES = Object.freeze(new Set(["audio", "music"]));
export const MULTIMODAL_CATEGORIES = Object.freeze(new Set(["image", "audio", "music", "video"]));

export const INTENT_MODALITY = Object.freeze({
  chat: "text",
  code: "text",
  image: "image",
  audio: "audio",
  music: "audio",
  video: "video",
});

const MUSIC_RE = /musicgen|jasco|ace-?step|acestep|diffrhythm|stable[\s_-]?audio|\bmagnet\b|magnet-|yue2|\byue\b|text-to-music|text2music/i;
const AUDIO_RE = /whisper|wav2vec|audiogen|faster-whisper|distil-whisper|speech[\s_-]?to[\s_-]?text|\bstt\b|\btts\b/i;
const CODE_RE = /coder|starcoder|codestral|code/i;

export function modalityOf(category) {
  return INTENT_MODALITY[category] ?? null;
}

export function isMultimodalCategory(category) {
  return MULTIMODAL_CATEGORIES.has(category);
}

export function modelCategory(displayName, hfId = "") {
  const text = `${displayName ?? ""} ${hfId ?? ""}`;
  if (MUSIC_RE.test(text)) return "music";
  if (AUDIO_RE.test(text)) return "audio";
  if (CODE_RE.test(text)) return "code";
  return "chat";
}

/** Sortable realtime factor (higher = faster). Music RTF (wall/audio) inverts to ×real. */
export function audioRealtimeValue(cell) {
  if (cell == null) return null;
  if (cell.audioXReal != null) return cell.audioXReal;
  if (cell.rtf != null && cell.rtf > 0) return 1 / cell.rtf;
  return null;
}

/** Multimodal metric path. Text cells (tok/s) return null — callers keep decode_tok_s. */
export function multimodalMetricOf(cell) {
  if (cell == null || cell.category == null) return null;
  if (cell.category === "image" && cell.imagesPerSec != null) {
    return { value: cell.imagesPerSec, unit: "img/s", label: "images" };
  }
  if (AUDIO_MODALITY_CATEGORIES.has(cell.category)) {
    const value = audioRealtimeValue(cell);
    if (value != null) return { value, unit: "×real", label: "realtime" };
  }
  if (cell.category === "video" && cell.videoFramesPerSec != null) {
    return { value: cell.videoFramesPerSec, unit: "f/s", label: "frames" };
  }
  return null;
}
